cloudconfig/machinecloudconfig.go

```diff
@@ -0,0 +1,291 @@
+	return machineCloudConfigV078
+}
+
+// machineCloudConfigV078 finds the containerInheritProperties properties and
+// values in the given dataMap and returns a cloud-init v0.7.8 formatted map.
+func machineCloudConfigV078(containerInheritProperties string, dataMap map[string]interface{}, log loggo.Logger) map[string]interface{} {
+	if containerInheritProperties == "" {
+		return nil
+	}
+	foundDataMap := make(map[string]interface{})
+	for _, k := range strings.Split(containerInheritProperties, ",") {
+		key := strings.TrimSpace(k)
+		switch key {
+		case "apt-security", "apt-sources", "apt-primary":
+			if val, ok := dataMap["apt"]; ok {
+				for k, v := range machineCloudConfigAptV078(key, val, log) {
+					// security, sources, and primary all nest under apt, ensure
+					// we don't overwrite prior translated data.
+					if apt, ok := foundDataMap["apt"].(map[string]interface{}); ok {
+						apt[k] = v
+					} else {
+						foundDataMap["apt"] = map[string]interface{}{
+							k: v,
+						}
+					}
+				}
+			} else {
+				log.Debugf("%s not found in machine cloud-init data", key)
+			}
+		case "ca-certs":
+			// no translation needed, ca-certs the same in both versions of cloudinit
+			if val, ok := dataMap[key]; ok {
+				foundDataMap[key] = val
+			} else {
+				log.Debugf("%s not found in machine cloud-init data", key)
+			}
+		}
+	}
+	return foundDataMap
+}
+
+func machineCloudConfigAptV078(key string, val interface{}, log loggo.Logger) map[string]interface{} {
+	split := strings.Split(key, "-")
+	secondary := split[1]
+
+	for k, v := range interfaceToMapStringInterface(val) {
+		if k == secondary {
+			foundDataMap := make(map[string]interface{})
+			foundDataMap[k] = v
+			return foundDataMap
+		}
+	}
+
+	log.Debugf("%s not found in machine cloud-init data", key)
+	return nil
+}
+
+var aptPrimaryKeys = []string{"apt_mirror", "apt_mirror_search", "apt_mirror_search_dns"}
+var aptSourcesKeys = []string{"apt_sources"}
+
+// machineCloudConfigV077 finds the containerInheritProperties properties and
+// values in the given dataMap and returns a cloud-init v0.7.7 formatted map.
+func machineCloudConfigV077(containerInheritProperties string, dataMap map[string]interface{}, log loggo.Logger) map[string]interface{} {
+	if containerInheritProperties == "" {
+		return nil
+	}
+	foundDataMap := make(map[string]interface{})
+	keySplit := strings.Split(containerInheritProperties, ",")
+	for _, k := range keySplit {
+		key := strings.TrimSpace(k)
+		switch key {
+		case "apt-primary", "apt-sources":
+			for _, aptKey := range append(aptPrimaryKeys, aptSourcesKeys...) {
+				if val, ok := dataMap[aptKey]; ok {
+					foundDataMap[aptKey] = val
+				} else {
+					log.Debugf("%s not found as part of %s, in machine cloud-init data", strings.Join(keySplit, "-"), key)
+				}
+			}
+		case "apt-security":
+			// Translation for apt-security unknown at this time.
+			log.Debugf("%s not found in machine cloud-init data", key)
```

Comment:
should this be a warning? or return an error? (if the user ha specified something that can't be fulfilled)

Reply:
changed to return an error.